bin/propolis-server/src/lib/vm/request_queue.rs

```diff
@@ -164,248 +214,330 @@ enum RequestDisposition {
+    /// the error contains a [`RequestDeniedReason`] that describes why the
+    /// request was rejected.
     pub fn try_queue(
         &mut self,
         request: ExternalRequest,
     ) -> Result<(), RequestDeniedReason> {
-        let disposition = match request {
-            ExternalRequest::Start => self.allowed.start,
-            ExternalRequest::MigrateAsSource { .. } => {
-                self.allowed.migrate_as_source
-            }
-            ExternalRequest::Reboot => self.allowed.reboot,
-            ExternalRequest::ReconfigureCrucibleVolume { .. } => {
-                self.allowed.mutate
+        // If the queue is in a terminal state, deny the request straightaway
+        // (unless it's a stop request, which can be ignored for idempotency).
+        let disposition = if let Some(reason) = self.state.deny_reason() {
+            if matches!(
+                request,
+                ExternalRequest::State(StateChangeRequest::Stop)
+            ) {
+                Disposition::Ignore
+            } else {
+                Disposition::Deny(reason)
             }
+        } else {
+            // The instance hasn't stopped yet, so consider this request in
+            // light of its current state and the other as-yet unprocessed
+            // requests from the queue.
+            //
+            // In general, try to make state change requests idempotent by
+            // ignoring new requests when a request of the appropriate kind is
+            // already on the queue, and deny requests to reach a state that is
+            // precluded by an earlier state change request.
+            match request {
+                // Interpret start requests as requests to reach the Running
+                // state.
+                ExternalRequest::State(StateChangeRequest::Start) => {
+                    if self.awaiting_stop {
+                        Disposition::Deny(RequestDeniedReason::HaltPending)
+                    } else if let QueueState::NotStarted = self.state {
+                        Disposition::Enqueue
+                    } else {
+                        Disposition::Ignore
+                    }
+                }
+
+                // Only allow one attempt to migrate out at a time (if it works
+                // the VM can't migrate out again), and only allow migration out
+                // after an instance begins to run.
+                ExternalRequest::State(
+                    StateChangeRequest::MigrateAsSource { .. },
+                ) => {
+                    if self.awaiting_migration_out {
+                        Disposition::Deny(
+                            RequestDeniedReason::AlreadyMigrationSource,
+                        )
+                    } else if self.awaiting_stop {
+                        Disposition::Deny(RequestDeniedReason::HaltPending)
+                    } else if matches!(self.state, QueueState::NotStarted) {
```

Comment:
nit: seems `QueueState` could also derive `PartialEq`...or, we could use `if let` here as we do above? but i'd just derive `PartialEq`...